File: src/api/evaluations.ts
import { RequestFN } from "~/core/request-function";
import { decodeEvaluation } from "~/decoders/evaluation";
import { encodePeriod } from "~/encoders/period";
import { type Evaluation, type Period, type SessionHandle, TabLocation } from "~/models";

export const evaluations = async (session: SessionHandle, period: Period): Promise<Array<Evaluation>> => {
  const request = new RequestFN(session, "DernieresEvaluations", {
    _Signature_: { onglet: TabLocation.Evaluations },

    donnees: {
      periode: encodePeriod(period)
    }
  });

  const response = await request.send();

  return response.data.donnees.listeEvaluations.V
    .map(decodeEvaluation);
};

File: library/src/commonMain/kotlin/api/evaluations.kt
package api

import core.RequestFN
import decoders.decodeEvaluation
import encoders.encodePeriod
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import models.Evaluation
import models.Period
import models.SessionHandle
import models.TabLocation

suspend fun evaluations (session: SessionHandle, period: Period): List<Evaluation> {
    val request = RequestFN(session.information, "DernieresEvaluations", Json.encodeToString(buildJsonObject {
        putJsonObject("_Signature_") {
            put("onglet", TabLocation.Evaluations.code)
        }

        putJsonObject("donnees") {
            put("periode", encodePeriod(period))
        }
    }))

    val response = request.send()

    return Json.parseToJsonElement(response.data).jsonObject["donnees"]!!.jsonObject["listeEvaluations"]!!.jsonObject["V"]!!.jsonArray.map { decodeEvaluation(it.jsonObject) }
}